### PR TITLE
Allow update of parameters.

### PR DESCRIPTION
**Pull request by trivialfis**

### CODE REVIEW

**Comment by trivialfis** [comment]
@hcho3 @RAMitchell Will give a demo in XGBoost after being ready.

**Comment by trivialfis** [comment]
ping @hcho3

File: test/unittest/unittest_param.cc [hunk 1]
```diff
@@ -20,7 +21,8 @@ TEST(Parameter, parsing_float) {
   std::map<std::string, std::string> kwargs;
 
   kwargs["float_param"] = "0";
-  ASSERT_NO_THROW(param.Init(kwargs));
+  param.Init(kwargs);
+  // ASSERT_NO_THROW(param.Init(kwargs));
```
Comment (hcho3) [inline test/unittest/unittest_param.cc]: Why are we commenting out this line?

Comment (trivialfis) [reply test/unittest/unittest_param.cc]: Oops.

File: test/unittest/unittest_param.cc [hunk 2]
```diff
@@ -157,3 +159,30 @@ TEST(Parameter, parsing_float) {
...
+  param.float_param = 0.02;
+  param.UpdateAllowUnknown(Args{{"float_param", "0.02"},
+                                {"foo", "bar"}}, &changed);
+  ASSERT_FALSE(changed);
+
+  param.UpdateAllowUnknown(Args{{"foo", "bar"}}, &changed);
+  ASSERT_FALSE(changed);
+
+  param.UpdateAllowUnknown(Args{{"double_param", "0.13"},
+                                {"foo", "bar"}}, &changed);
```
Comment (hcho3) [inline test/unittest/unittest_param.cc]: What is the expected value of `changed` after this statement?

**Comment by trivialfis** [comment]
Fixed.

File: include/dmlc/parameter.h [hunk 3]
```diff
@@ -575,12 +642,23 @@ class FieldEntryBase : public FieldAccessEntry {
...
-  virtual std::string GetStringValue(void *head) const {
+  bool Same(void* head, std::string const& value) const override {
+    DType old = this->Get(head);
+    DType now;
+    std::istringstream is(value);
+    is >> now;
+    // don't require = operator
+    bool is_same = std::equal(
+        reinterpret_cast<char*>(&now), reinterpret_cast<char*>(&now) + sizeof(now),
+        reinterpret_cast<char*>(&old));
```
Comment (hcho3) [inline include/dmlc/parameter.h]: Are `old` and `new` same only when `old` and `new` have same set of bits? How about `old=-0.0` and `new=0.0`? What is the reason for avoiding the use of `=` operator?

Comment (hcho3) [reply include/dmlc/parameter.h]: According to IEEE 754, `-0.0` and `0.0` are distinct values (with different bits) and at the same time `-0.0 == 0.0` should evaluate to `true`.

**Comment by trivialfis** [comment]
There's a vector in XGB tree param.

We might be able to use type traits to test the concept `EqualityComparable` defined in C++ 20.  Let me try that.

**Comment by hcho3** [comment]
Got it. For now, bit equality should be good.

**Comment by trivialfis** [comment]
@hcho3 Can we merge it?  I will update `dmlc-core` in XGBoost.

**Pull request merged**